src/vs/vscode.proposed.d.ts

```diff
@@ -1554,24 +1564,76 @@ declare module 'vscode' {
 	}
 
 	export interface NotebookProvider {
-		resolveNotebook(editor: NotebookEditor): Promise<void>;
-		executeCell(document: NotebookDocument, cell: NotebookCell | undefined): Promise<void>;
+		resolveNotebook(editor: NotebookEditor): Promise<void>; // rchiodo: Not sure what this is for? Notebook is loaded? Like prime the server? Or does this somehow generate the document?
 		save(document: NotebookDocument): Promise<boolean>;
 	}
 
+	/**
+	 * Result of an executeCell.
+	 */
+	export interface NotebookCellExecutionResult {
+		/**
+		 * Event fired when new cell output comes through
+		 */
+		readonly onCellOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when output should be cleared. Boolean flag indicates if should wait for next onOutput for clear.
+		 */
+		readonly onClear: Event< boolean>;
+		/**
+		 * Event fired when document wide output comes through
+		 */
+		readonly onDocumentOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when execution is complete
+		 */
+		readonly onComplete: Event<void>;
+	}
+
+	/**
+	 * Allows execution of notebook cells.
+	 */
+	export interface NotebookExecution {
+		readonly executionCount: number;
+		executeCell(document: NotebookDocument, cell: NotebookCell): NotebookCellExecutionResult;
+	}
+
+	/**
+	 * Allows observing the execution of cells from a NotebookExecution
+	 */
+	export interface NotebookExecutionObserver {
+		onExecuteStart(document: NotebookDocument, cell: NotebookCell): void;
+		onCellOutput(document: NotebookDocument, cell: NotebookCell, outputs: CellOutput[]): void;
+		onCellClear(document: NotebookDocument, cell: NotebookCell, wait: boolean): void;
+		onDocumentOutput(document: NotebookDocument, cell: NotebookCell, outputs: CellOutput[]): void;
+		onExecuteComplete(document: NotebookDocument, cell: NotebookCell, executionCount: number): void;
```

Comment:
Not sure about this, now that we have `metadata` on each cell, why do we need these?

Ie. why would one need `onExecutionStart` and `onExecuteionComplete`.
Does an extension need to know something has started? I would have expected an extension to merely know execution completed, and that information cab be derived from the fact that something was rendered for a cell.

I'm thinking of `gather`, we only need to know when something has completed.
Similarly `cellClear` that can be communicated with `onCellOutput` having empty `outputs`.

Reply:
OnCellClear is a result of the jupyter message - clear a cell. You need to have this in order to know when to clear (might be now, might be after). 

The Observer doesn't have the underlying document, it just has the events. There'd be no onCellOutput for actually clearing the output. 

onExecutionStart is necessary so you can kick off your own execution (like the variable explorer would need)

Reply:
> OnCellClear is a result of the jupyter message - clear a cell. You need to have this in order to know when to clear (might be now, might be after).

As mentioned earlier, this can be achieved by telling VSC to render and empty result. Empty result = clear output.

> onExecutionStart is necessary so you can kick off your own execution (like the variable explorer would need)

Hmm, I'd expect this to only come up `after` execution, not while.
I.e. cell has been executed, now lets display the variable explorer.

Reply:
You don't tell VSC to render. I think you're missing the mental model here.

There's a NotebookDocument. It has cells in it. VS code owns this object. When a user clicks 'run cell', VS code finds the NotebookExecution and calls executeCell on it. It then signs up for the events in the NotebookExecutionResult. 

onClear comes in as an event. VS code must then clear the cell outputs in its representation of the Document. There are no extensions that change the Document's outputs. VS code does that.

Reply:
As far as the variable explorer, it needs to execute the same code as the real kernel so that it's internal state is the same. That's why it needs to know when the cell begins.

Reply:
Why not do this:

* VSC fires events when buttons are clicked (fire commands) as they do today
* Extensions do their stuff and send results to VSC

The only thing thats missing is `trigger for extension to start executing`. I.e. if we looked at the existing model (what VSC have today), VSC does not have a concept of execution it only renders stuff. It handles buttons clicks like `ADD CELL`,etc..

I think we can still do this by just letting VSC expose events for `executeCell` or similar, I.e.g `Command` with a cell.
This way we can have commands for `Clear Cell`, `Run Cell`, etc.. 

If feel that fits with little change to existing API, and gives us what we (extension authors) need.

Reply:
I think you're over simplifying. The Command for 'execute' cell is the same as the API here. And it doesn't allow other extensions to listen to the command. Commands are single listener at a time.

And how does this part work?
> Extensions do their stuff and send results to VSC

How do they send results? The results aren't a thenable action. They're a multilevel set of messages that come back. Hence the NotebookExecutionResult. I'd much rather use an observable, but they don't support that in VS code.

Reply:
And I think the clear I'm talking about is not the same clear you're thinking of. My clear comes from this:

```
from IPython.display import clear_output

for i in range(10):
    clear_output()
    print("Hello World!")
```

Not a button on a cell. It's stuck in the middle of outputting.

Reply:
> And it doesn't allow other extensions to listen to the command. Commands are single listener at a time.

Again, the only piece missing is triggering extension to execute cells. Should we first try to solve that trhan adding concept of execution and the like. 
I.e. solve problem of VSC sending UI events to extension.

Reply:
> Hence the NotebookExecutionResult. I'd much rather use an observable, but they don't support that in VS code.

Why does VSC need observable. Its extension that needs observable to update the ui.
In react we don't have observable, we merely render the data as and when required. I.e. observable is merely used to `react` to sometning that happened. In this case, the result of the `reaction` is to render a result.

Reply:
The execution needs the observable because it's a mulitstep thing. Rendering doesn't.

I don't think discussing this here is working. I'll setup a meeting on Monday